src/cargo/sources/registry/http_remote.rs

```diff
@@ -0,0 +1,666 @@
+        } else if self.config.cli_unstable().no_index_update {
+            trace!("using local {} in no_index_update mode", path.display());
+            true
+        } else if self.config.offline() {
+            trace!("using local {} in offline mode", path.display());
+            true
+        } else if self.fresh.contains(path) {
+            trace!("using local {} as it was already fetched", path.display());
+            true
+        } else {
+            debug!("checking freshness of {}", path.display());
+            false
+        }
+    }
+}
+
+impl<'cfg> RegistryData for HttpRegistry<'cfg> {
+    fn prepare(&self) -> CargoResult<()> {
+        Ok(())
+    }
+
+    fn index_path(&self) -> &Filesystem {
+        &self.index_path
+    }
+
+    fn assert_index_locked<'a>(&self, path: &'a Filesystem) -> &'a Path {
+        self.config.assert_package_cache_locked(path)
+    }
+
+    fn is_updated(&self) -> bool {
+        self.requested_update
+    }
+
+    fn load(
+        &mut self,
+        _root: &Path,
+        path: &Path,
+        index_version: Option<&str>,
+    ) -> Poll<CargoResult<LoadResponse>> {
+        trace!("load: {}", path.display());
+        if let Some(_token) = self.downloads.pending_ids.get(path) {
+            debug!("dependency is still pending: {}", path.display());
+            return Poll::Pending;
+        }
+
+        if let Some(index_version) = index_version {
+            trace!(
+                "local cache of {} is available at version `{}`",
+                path.display(),
+                index_version
+            );
+            if self.is_fresh(path) {
+                return Poll::Ready(Ok(LoadResponse::CacheValid));
+            }
+        } else if self.fresh.contains(path) {
+            debug!(
+                "cache did not contain previously downloaded file {}",
+                path.display()
+            );
+        }
+
+        if let Some(result) = self.downloads.results.remove(path) {
+            let result =
+                result.with_context(|| format!("download of {} failed", path.display()))?;
+            debug!(
+                "index file downloaded with status code {}",
+                result.response_code
+            );
+            trace!("index file version: {}", result.index_version);
+
+            if !self.fresh.insert(path.to_path_buf()) {
+                debug!("downloaded the index file `{}` twice", path.display())
+            }
+
+            match result.response_code {
+                200 => {}
+                304 => {
+                    // Not Modified: the data in the cache is still the latest.
+                    if index_version.is_none() {
+                        return Poll::Ready(Err(anyhow::anyhow!(
+                            "server said not modified (HTTP 304) when no local cache exists"
+                        )));
+                    }
+                    return Poll::Ready(Ok(LoadResponse::CacheValid));
+                }
```

Comment:
Should this handle HTTP redirects? That's a design question that was not thought about in the RFC.

Not a blocker just something I thought of while reviewing.

Reply:
It currently does follow redirects. It calls `handle.follow_location(true)?;` to enable it.

Of course, performance will be worse since it has to do 2x the requests - but that's up to the registry operator.

Reply:
Good, glad to here that it is Curls problem.